setup.py

```diff
@@ -22,30 +23,16 @@ def _read_me() -> str:
     return readme
 
 
-def _get_version() -> str:
-    # ops.__init__ needs to pull in ops.charm to work around a circular
-    # import so we can't import it here as that pulls in yaml which isn't
-    # necessarily there yet.
-    version = 'unknown'
-    with open("ops/__init__.py", "rt", encoding="utf8") as fh:
-        source = fh.read()
-    code = ast.parse(source)
-    for node in code.body:
-        if isinstance(node, ast.Assign):
-            targets = [i.id for i in node.targets]
-            if '__version__' in targets:
-                if isinstance(node.value, ast.Str):
-                    # Python < 3.8
-                    version = node.value.s
-                else:
-                    version = node.value.value
-                break
-    return version
-
+version = importlib.import_module('ops.version').version
```

Comment:
Doesn't this still trigger the bad code path of importing version forces us to import __init__ which grabs YAML before have finished running setup.py to install it?

Reply:
nope :-)

Reply:
It looks you can't do that:

```
** fades ***  2020-06-04 16:02:53,040  fades.pipmanager   INFO     Installing dependency: 'file:///home/facundo/canonical/operator'
*** fades ***  2020-06-04 16:02:53,040  fades.exec         DEBUG    Executing external command: ['/home/facundo/.local/share/fades/580857b5-9a06-4353-9c7d-1a71d054c356/bin/pip', 'install', 'file:///home/facundo/canonical/operator']
*** fades ***  2020-06-04 16:02:53,860  fades.exec         DEBUG    :: Processing ./canonical/operator
*** fades ***  2020-06-04 16:03:16,726  fades.exec         DEBUG    ::     Complete output from command python setup.py egg_info:
*** fades ***  2020-06-04 16:03:16,726  fades.exec         DEBUG    ::     Traceback (most recent call last):
*** fades ***  2020-06-04 16:03:16,726  fades.exec         DEBUG    ::       File "<string>", line 1, in <module>
*** fades ***  2020-06-04 16:03:16,726  fades.exec         DEBUG    ::       File "/tmp/pip-plgmdasp-build/setup.py", line 26, in <module>
*** fades ***  2020-06-04 16:03:16,726  fades.exec         DEBUG    ::         version = importlib.import_module('ops.version').version
*** fades ***  2020-06-04 16:03:16,727  fades.exec         DEBUG    ::       File "/usr/lib/python3.6/importlib/__init__.py", line 126, in import_module
*** fades ***  2020-06-04 16:03:16,727  fades.exec         DEBUG    ::         return _bootstrap._gcd_import(name[level:], package, level)
*** fades ***  2020-06-04 16:03:16,727  fades.exec         DEBUG    ::       File "<frozen importlib._bootstrap>", line 994, in _gcd_import
*** fades ***  2020-06-04 16:03:16,727  fades.exec         DEBUG    ::       File "<frozen importlib._bootstrap>", line 971, in _find_and_load
*** fades ***  2020-06-04 16:03:16,727  fades.exec         DEBUG    ::       File "<frozen importlib._bootstrap>", line 941, in _find_and_load_unlocked
*** fades ***  2020-06-04 16:03:16,727  fades.exec         DEBUG    ::       File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
*** fades ***  2020-06-04 16:03:16,727  fades.exec         DEBUG    ::       File "<frozen importlib._bootstrap>", line 994, in _gcd_import
*** fades ***  2020-06-04 16:03:16,727  fades.exec         DEBUG    ::       File "<frozen importlib._bootstrap>", line 971, in _find_and_load
*** fades ***  2020-06-04 16:03:16,727  fades.exec         DEBUG    ::       File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
*** fades ***  2020-06-04 16:03:16,727  fades.exec         DEBUG    ::       File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
*** fades ***  2020-06-04 16:03:16,727  fades.exec         DEBUG    ::       File "<frozen importlib._bootstrap_external>", line 678, in exec_module
*** fades ***  2020-06-04 16:03:16,728  fades.exec         DEBUG    ::       File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
*** fades ***  2020-06-04 16:03:16,728  fades.exec         DEBUG    ::       File "/tmp/pip-plgmdasp-build/ops/__init__.py", line 20, in <module>
*** fades ***  2020-06-04 16:03:16,728  fades.exec         DEBUG    ::         from . import charm  # noqa: F401 (imported but unused)
*** fades ***  2020-06-04 16:03:16,728  fades.exec         DEBUG    ::       File "/tmp/pip-plgmdasp-build/ops/charm.py", line 20, in <module>
*** fades ***  2020-06-04 16:03:16,728  fades.exec         DEBUG    ::         import yaml
*** fades ***  2020-06-04 16:03:16,728  fades.exec         DEBUG    ::     ModuleNotFoundError: No module named 'yaml'
*** fades ***  2020-06-04 16:03:16,728  fades.exec         DEBUG    ::     
*** fades ***  2020-06-04 16:03:16,728  fades.exec         DEBUG    ::     ----------------------------------------
*** fades ***  2020-06-04 16:03:16,728  fades.exec         DEBUG    :: Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-plgmdasp-build/
```

Reply:
that's strange. I'll dig a bit.

Reply:
> Doesn't this still trigger the bad code path of importing version forces us to import **init** which grabs YAML before have finished running setup.py to install it?

I swear I tested this, running with `-v` to see the imports, and it wasn't loading `__init__.py`. But obviously I was doing something wrong, as @facundobatista's quick test showed.

The code that actually does what I want is a little bit longer so I moved it to a function, but now it works.